File: java/src/test/java/org/oojs/AdditionalPropertiesTest.java
package org.oojs;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class AdditionalPropertiesTest extends TestHelpers {

    private Registry r;
    private Schema schema;

    @BeforeEach void setUp() {
        Object[] ar = animalRegistry();
        r = (Registry) ar[0];
        schema = (Schema) ar[1];
    }

    @Test void additional_property_rejected() {
        List<ValidationError> errs = new Validator(r).validate(
                map("_type", "Dog", "name", "Rex", "breed", "Lab", "color", "black"),
                schema.types.get("Dog"), schema);
        assertTrue(hasCode(errs, ErrorCode.ADDITIONAL_PROPERTY));
    }

    @Test void open_world_allows_extra() {
        Registry r2 = new Registry();
        r2.loadMap(map(
                "$oojs", "1.0", "$id", "https://example.org/schemas/open",
                "additionalProperties", true,
                "types", map("Foo", map("properties", map("x", map("type", "string"))))));
        Schema s = r2.getSchema("https://example.org/schemas/open");
        List<ValidationError> errs = new Validator(r2).validate(
                map("_type", "Foo", "x", "hello", "extra", 99), s.types.get("Foo"), s);
        assertEquals(List.of(), errs);
    }
}

File: php/tests/AdditionalPropertiesTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\ErrorCode;
use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * §8.4, §9.2 Phase 3 — Closed-world / additional properties.
 */
class AdditionalPropertiesTest extends TestCase
{
    use TestHelpers;

    private Registry $r;
    private \Oojs\Model\Schema $schema;

    protected function setUp(): void
    {
        [$this->r, $this->schema] = self::animalRegistry();
    }

    public function test_additional_property_rejected(): void
    {
        $errs = validate(
            ['_type' => 'Dog', 'name' => 'Rex', 'breed' => 'Lab', 'color' => 'black'],
            $this->schema->types['Dog'],
            $this->schema,
            $this->r,
        );
        $this->assertTrue($this->hasCode($errs, ErrorCode::ADDITIONAL_PROPERTY));
    }

    public function test_open_world_allows_extra(): void
    {
        $r = new Registry();
        $r->loadDict([
            '$oojs'                => '1.0',
            '$id'                  => 'https://example.org/schemas/open',
            'additionalProperties' => true,
            'types'                => ['Foo' => ['properties' => ['x' => ['type' => 'string']]]],
        ]);
        $schema = $r->getSchema('https://example.org/schemas/open');
        $errs   = validate(
            ['_type' => 'Foo', 'x' => 'hello', 'extra' => 99],
            $schema->types['Foo'],
            $schema,
            $r,
        );
        $this->assertSame([], $errs);
    }
}
